test/persistence-hooks.suite.js

```diff
@@ -2946,6 +2946,252 @@ module.exports = function(dataSource, should, connectorCapabilities) {
 
     });
 
+    describe('PersistedModel.upsertWithWhere', function() {
+      var createdId = null;
+      var createdExtra = null;
+      var createdName = null;
+
+      it('triggers hooks in the correct order on create', function(done) {
+        monitorHookExecution();
+        TestModel.upsertWithWhere({ where: { extra: 'e0101' }},
+          { id: 'not-found', name: 'not found' },
+          function(err, record, created) {
+            if (err) return done(err);
+            createdId = record.id;
+            createdExtra = record.extra;
+            createdName = record.name;
+            hookMonitor.names.should.eql([
+              'access',
+              'before save',
+              'persist',
+              'loaded',
+              'after save',
+            ]);
+            done();
+          });
+      });
+
+      it('triggers hooks in the correct order on update', function(done) {
+        monitorHookExecution();
+        TestModel.upsertWithWhere({ where: { extra: 'e0101' }},
+          { name: 'new name' },
+          function(err, record, created) {
+            if (err) return done(err);
+            hookMonitor.names.should.eql([
+              'access',
+              'before save',
+              'persist',
+              'loaded',
+              'after save',
+            ]);
+            done();
+          });
+      });
```

Comment:
This test case is not complete; you should check the order of `hooks` for two scenarios:
* When we have the atomic implementation of your method `upsertWithWhere`
* When we **DO NOT** have the atomic implementation of your method `upsertWithWhere`

The order of hooks would be different since non atomic implementation calls `find()` and `create()`/`updateAttributes()` where they have their own hooks, whereas the order of `hooks` for atomic implementation would be different; you should check it something like:

```
            if (dataSource.connector.upsertWithWhere) {
              hookMonitor.names.should.eql([
                ...
              ]);
            } else {
              hookMonitor.names.should.eql([
                ...
              ]);
            }
```